src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -0,0 +1,234 @@
+                           << enums.size() << std::endl;
+  NodeManager* nm = NodeManager::currentNM();
+  for (unsigned i = 0, ecsize = enum_consider.size(); i < ecsize; ++i)
+  {
+    unsigned j = enum_consider[i];
+    Node e = enums[j];
+    Node v = enum_values[j];
+    std::vector<Node> enum_lems;
+    d_sygus_unif.notifyEnumeration(e, v, enum_lems);
+    /* the lemmas must be guarded by the active guard of the enumerator */
+    Assert(d_enum_to_active_guard.find(e) != d_enum_to_active_guard.end());
+    Node g = d_enum_to_active_guard[e];
+    for (unsigned j = 0, size = enum_lems.size(); j < size; ++j)
+    {
+      enum_lems[j] = nm->mkNode(OR, g.negate(), enum_lems[j]);
+    }
+    lems.insert(lems.end(), enum_lems.begin(), enum_lems.end());
+  }
+  /* build candidate solution */
+  Assert(candidates.size() == 1);
+  Node vc = d_sygus_unif.constructSolution();
+  if (vc.isNull())
+  {
+    return false;
+  }
+  candidate_values.push_back(vc);
+  return true;
+}
+
+Node CegisUnif::purifyLemma(Node n,
+                            bool ensureConst,
+                            std::vector<Node>& model_guards,
+                            BoolNodePairMap& cache)
+{
+  Trace("cegis-unif-purify") << "PurifyLemma : " << n << "\n";
+  BoolNodePairMap::const_iterator it = cache.find(BoolNodePair(ensureConst, n));
+  if (it != cache.end())
+  {
+    Trace("cegis-unif-purify") << "... already visited " << n << "\n";
+    return it->second;
+  }
+  /* Recurse */
+  unsigned size = n.getNumChildren();
+  Kind k = n.getKind();
+  bool fapp = k == APPLY_UF && size > 0 && n[0] == d_candidate;
+  bool childChanged = false;
+  std::vector<Node> children;
+  for (unsigned i = 0; i < size; ++i)
+  {
+    Node child = purifyLemma(n[i], ensureConst || fapp, model_guards, cache);
+    children.push_back(child);
+    childChanged = childChanged || child != n[i];
+  }
+  Node nb;
+  if (childChanged)
+  {
+    if (n.hasOperator())
+    {
+      children.insert(children.begin(), n.getOperator());
+    }
+    nb = NodeManager::currentNM()->mkNode(k, children);
+    Trace("cegis-unif-purify") << "PurifyLemma : transformed " << n << " into "
+                               << nb << "\n";
+  }
+  else
+  {
+    nb = n;
+  }
+  /* get model value of non-top level applications of d_candidate */
+  if (ensureConst && fapp)
+  {
+    Node nv = d_parent->getModelValue(nb);
+    Trace("cegis-unif-purify") << "PurifyLemma : model value for " << nb
+                               << " is " << nv << "\n";
+    /* test if different, then update model_guards */
+    if (nv != nb)
+    {
+      Trace("cegis-unif-purify") << "PurifyLemma : adding model eq\n";
+      model_guards.push_back(
+          NodeManager::currentNM()->mkNode(EQUAL, nv, nb).negate());
+      nb = nv;
+    }
+  }
+  /* every non-top level application of function-to-synthesize must be reduced
+     to a concrete constant */
+  Assert(!ensureConst || !fapp || nb.isConst());
```

Comment:
I agree we should rewrite ``nb`` however if I remove ``!fapp`` from the assertion then I have to test than ``nb == d_candidate`` because we do not rewrite the deep embedding of the function into the function (its model value), and we want to keep the deep embedding.